internetarchive/openlibrary
# Author merge requests

Closes #6812


Allows author merge requests to be added to the librarian request queue.  Author merge requests are added similarly to work merge requests: `librarians` can submit a request via an author search results page via the ILE blue bar, and `super-librarians` can accept or decline the request from the author merge UI.  Author merges by `super-librarians` trigger the creation of an approved merge request.

### Technical

This PR alters the `community_edits_queue` table, adding a new column 'type` which is used to differentiate work and author merge requests.  __Before this is deployed to any environment, the following queries should be executed:__
```
ALTER TABLE community_edits_queue ADD type int;
UPDATE community_edits_queue set type = 1; /* designates all existing requests as work merges */
```

The `/merges` POST controller has significantly changed.  MR `url` and `title` generation is now handled by the controller.  Total request types (`rtype`) accepted by the controller have been reduced to two: `create-request` and `update-request`.

The author merge flow seems a bit unusual to me.  Here is a refresher on how this works today:
1. The "Merge Authors" button in the author merge UI is clicked.
2. Data is POSTed to `/authors/merge`, and is validated
3. On successful validation, we are redirected to the parent author page.  The redirect URL contains `merge=true` and a list of duplicate OLIDs.  A loading indicator is displayed on the page (due to the `merge` query parameter's presence).
4. The loading indicator's presence on the page triggers a client-side request to the `/authors/merge` JSON endpoint. This triggers the author merge.
5. The loading indicator is updated when the JSON endpoint response is received.

We now pass an optional `mrid` and `comment` to the author page when a request is submitted.  If an `mrid` is passed to the JSON author merge endpoint, a MR is updated.  Otherwise, a new "Approved" author MR is created.

On author search result pages, `super-librarians` can access the author merge UI by using the usual "Merge authors" link (next to the magic wand near the top of the page), or by using the link on the ILE blue bar.  The old-school "Merge authors" link will display all of the authors in the merge UI, while specific authors can be displayed when the ILE merge link is used.

### Testing

Ensure that all work merge request flows still work as expected.
As a `librarian`:
1. Search for duplicate author candidates.  Select the duplicates and press the "Merge authors..." link in the ILE blue bar.
2. A dialog prompting for an optional comment will appear. Type a comment and press "OK".
3. Click on the link in the toast message to see the MR table.  Ensure that your MR is present, and that it is an author merge.
4. Add a comment on the author MR. Ensure that the view updates with the new comment.  Refresh the page, and ensure that the new comment is visible on page load.
5. Try closing the MR from the table view.  Click the "Closed" tab, and ensure that the newly closed MR is present.
6. Make two more author MRs.  These will be used to test author MR approval and rejection.
7. Switch to a `super-librarian` account and navigate to the MR table.
8. Enter the author merge UI by clicking the "merge" link for one of the author MRs.
9. Leave a comment and press the "Reject" button.  Ensure that the MR is in the "Closed" merge table view.
10. Switch back to the "Open" merge request table view, and click the "merge" link for the other author MR.
11. Select the records you'd like to merge, type a comment, and click the "Merge Authors" button.
12. You will be redirected to the author page of the parent record.  Once you see a success message, ensure that the MR is listed in the "Closed" view as "Merged".
13. Finally, test the whole author merge flow as a `super-librarian`.  Select authors from a search result page and click the merge link in the ILE blue bar.  Note that the "Reject" button has been replaced by the "Cancel" link.  Select the authors and press the merge button.  On success, ensure that a new entry was created in the "Closed" MR table view.

### Screenshot

![Screenshot from 2022-08-11 15-59-26](https://user-images.githubusercontent.com/28732543/184229686-c5d92083-c2c8-4f5c-a67c-a7724d0a5951.png)
_New comment input and rejection button_

![Screenshot from 2022-08-11 16-03-13](https://user-images.githubusercontent.com/28732543/184230494-b45418b6-5129-4d13-b165-1039ca79f901.png)
_Merge types now present in MR titles_

### Stakeholders
